### DISCUSSION TITLE
How to get the in sample predictions

### DISCUSSION THREAD

**Discussion by dirknbr** [alan-turing-institute/sktime#2126] [Q&A]
I would like to get the in sample predictions of a forecaster, to estimate residuals and e**2 

It seems fh always has to be positive and in the future

How can I use .predict() for this?

Maybe only some forecasters support it.

**Comment by aiwalter** [comment]
Just give negative values or insample index values for absolute fh. Also you might check older issues by searching for "residual"

**Comment by dirknbr** [comment]
Are you sure?

```
s = STLForecaster(sp=7)
s.fit(data.y[train])
s.predict(fh=[-2, -1])
....
ValueError: cannot reshape array of size 718 into shape (103,7)

```

I could not find other 'residual' threads.

**Reply by aiwalter** [reply]
Which version is this?

**Reply by aiwalter** [reply]
Try another forecaster, might be specific issue here

**Reply by aiwalter** [reply]
@eenticott-shell can you pls look into this? Looks like the bug from `NaiveForecaster` wasnt fully fixed?

**Reply by fkiraly** [reply]
Do you have a PR number, @aiwalter or @eenticott-shell ?

**Reply by fkiraly** [reply]
@aiwalter, what´s your reasoning behind the attribution to `NaiveForecaster`?

**Reply by fkiraly** [reply]
& why do the tests not catch this? We are testing in-sample forecasts, are we not?

**Reply by aiwalter** [reply]
@dirknbr I tried to reproduce your error: 
```python
from sktime.forecasting.trend import STLForecaster
from sktime.datasets import load_airline

y = load_airline()
s = STLForecaster(sp=7)
s.fit(y)
s.predict(fh=[-2, -1])
```

but couldnt get it, I am on 0.10.1 version

**Reply by aiwalter** [reply]
can you tell me what kind of data you fed in? Was it `numpy` or `pandas`? Pls provide example code

**Reply by fkiraly** [reply]
yes, let's treat this like a bug report.

We need:
* minimally reproducible code (incl all imports)
* version number of `sktime` and key dependencies

**Reply by fkiraly** [reply]
@dirknbr?

**Comment by fkiraly** [comment]
@dirknbr, according to interface specification `predict` should work with `fh` that is in-sample, e.g., absolute time stamps, or negative integers.
If that doesn't work and doesn't raise an informative error message ("this forecaster can't do that"), it's a bug.

Btw, there's a convenience function for in-sample residuals: `predict_residuals`!

**Comment by dirknbr** [comment]
I think the insample prediction works now. We can close
